### PR TITLE
decode with inline pointers

### LINKED ISSUE #210: Pointer in inline structs returns error when reading from database
Hi
I encountered an issue while populating a struct with db data. My structs are as follows:
```
type Product struct {
    ID bson.ObjectId `json:"id" bson:"_id"`
    Name string `json:"name" bson:"name"`
    Meta `json:"meta" bson:"meta"`
}

type Meta {
    *Printer `json:",inline" bson:",inline"`
    *Monitor `json:",inline" bson:",inline"`
}

type Monitor struct {
    ScreenSize float64 `json:"screen_size" bson:"screen_size"`
    ScreenType string `json:"screen_type" bson:"screen_type"`
}

...

p := Product{}
if err := db.C("products").Find(nil).All(&p); err != nil {
    return err
}

```
the error that I get from this action is
```
reflect: indirection through nil pointer to embedded struct
```
I should note that this only happens while reading from the db, and not when inserting into.

I am aware that this issue has been raised couple of times before: #146, go-mgo#346, and although a fix has been submitted, it still seems that it doesn't work properly (at least for reading data from the db).

The version of go that I'm using is
```
go version go1.10.3 darwin/amd64
```
and the globalsign/mgo version is the last one available

### PR DESCRIPTION
**Pull request by larrycinnabar**
fix #210

My last PR !146 added ability to encode bson with inline pointers to structs (as `json.Marshal` does)

Here, we add the ability to decode.
It's a little bit tricky.

 **help and/or review is appreciated**.

1) only exported inline fields are allowed to be decoded. (as `field.CanInterface()` will return false for protected inline fields)
Here couple of tests still needed - to check if proper error message is shown to user (that unexported inline pointers are not allowed)

2) there are 2 reflect helper created (to create a deep zero element) - that will be the base element when decoded. These functions need to be unit-test covered too.

Current behaviour might seem not correct. What do you think, is it ok?
```golang
type MStruct struct {
	M int `bson:"m,omitempty"`
}
type InlinePtrStruct struct {
	A        int
	*MStruct `bson:",inline"`
}

func (s *S) TestPtrInline(c *C) {
		in := InlinePtrStruct{A: 4}
		data, err := bson.Marshal(in)
		c.Assert(err, IsNil)

		out := InlinePtrStruct{}
		err = bson.Unmarshal(data, &out)
		c.Assert(err, IsNil)

		c.Assert(out.A, Equals, 4)
		c.Assert(out.M, Equals, 0)
}
```

This test passes. But after `Unmarshal` we get `out` to be: `InlinePtrStruct{A: 4, MStruct: &Mstruct{M:0}}`. So instead of `nil` inline structs we get pointers to empty structs. So, inline is always goes deep, and keeps pointers to empty structs.

### CODE REVIEW

**Comment by domodwyer** [comment]
Hi @larrycinnabar 

Thanks for the PR! Looking good so far.

Regarding the initialised `Mstruct` instance instead of `nil`, I think it would be preferred that it was indeed `nil` as a nil pointer is often used to convey meaning itself and we recently merged some not-totally-definitely-backwards-compatible [changes](https://github.com/globalsign/mgo/pull/147) to make it work as expected for slices and maps (thanks @aksentyev!) - do you think it's possible?

Again, thanks very much for take this on!

Dom

**Comment by larrycinnabar** [comment]
@domodwyer 

ok, right, thanks for note.

**Comment by larrycinnabar** [comment]
Fix is pushed. nil-respecting behavior is available through the respect-nil option.

Right before pushing I've found one more case, that should be handled. It's commented in tests. WIll handle it in 1-2 days.

File: bson/bson_test.go [hunk 1]
```diff
@@ -274,39 +274,125 @@ func (s *S) TestMarshalBuffer(c *C) {
...
+		err = bson.Unmarshal(data, &outRespectNils)
+		bson.SetRespectNilValues(false)
 		c.Assert(err, IsNil)
+		c.Assert(outRespectNils.InlineG1, IsNil)
+		c.Assert(outRespectNils, DeepEquals, in)
+	}
 
-		c.Assert(dataBSON, DeepEquals, cs.Out)
+	// mixing empty & nil
+	// @TODO
```
Comment (domodwyer) [inline bson/bson_test.go]: Is this just a forgotten test?

**Comment by domodwyer** [comment]
Hi @larrycinnabar 

**Really sorry for the delay** - it's been a really, really turbulent few of weeks!

I'll review over lunch today.

Dom

**Comment by domodwyer** [comment]
Looks great - just the commented test and it's good to go!

File: bson/bson_test.go (hunk 1 above)
Comment (larrycinnabar) [reply bson/bson_test.go]: Hm, I think it was a real TODO, not just forgotten test.
I had really turbulent few of weeks too, so I think I need to spend couple hours on this yet.

Let me look into it this weekend and then I'll push / or post here the update status

Comment (larrycinnabar) [reply bson/bson_test.go]: @domodwyer 

I've debugged the issue:

It's expected behaviour in the failing test, as `g1` field is `omitempty` it is lost while `marshalling`. That's why when we unmarshal back with `respect-nil` option - we lose it.

Removing `omitempty` from `g1` field from `InlineG1` struct makes test pass. 
So either remove `omitempty`, either remove that test.

**Comment by larrycinnabar** [comment]
I've debugged the issue:

It's expected behaviour in the failing test, as `g1` field is `omitempty` it is lost while `marshalling`. That's why when we `unmarshal` back with `respect-nil` option - we lose it.

Removing `omitempty` from `g1` field from `InlineG1` struct makes test pass.
Feel free to remove that test and merge then

**Comment by domodwyer** [comment]
Thanks @larrycinnabar! 

It's really appreciated :)

Dom

**Comment by eminano** [comment]
Thanks a lot for your contribution @larrycinnabar and sorry for the delay!

**Comment by msolimans** [comment]
@eminano thanks for the merge into dev branch! any plans for release?

**Comment by eminano** [comment]
Hi @msolimans, we'll be cutting a release soon, just need to get a bit of time to run all the performance tests required and we're set!

Esther

**Pull request merged**